File: tests_MATLAB/test5.m
% Poisson accuracy test

addpath('../mole_MATLAB')

west = 0;  % Domain's limits
east = 1;

ks = [2, 4, 6];  % Different orders of accuracy
grid_sizes = [20, 40];  % Grid sizes to test

for k = ks

    errors = zeros(size(grid_sizes));
    
    for i = 1:numel(grid_sizes)
        m = grid_sizes(i);  % Number of cells
        dx = (east - west) / m;  % Step length

        L = lap(k, m, dx);  % 1D Mimetic Laplacian operator

        % Impose Robin BC on Laplacian operator
        a = 1;
        b = 1;
        L = L + robinBC(k, m, dx, a, b);

        % 1D Staggered grid
        grid = [west west+dx/2 : dx : east-dx/2 east];

        % RHS
        U = exp(grid)';
        U(1) = 0;  % West BC
        U(end) = 2*exp(1);  % East BC

        % Solve a linear system of equations
        computed_solution = L\U;

        % Compute error using L2 norm
        analytical_solution = exp(grid);
        errors(i) = max(abs(computed_solution' - analytical_solution));
    end

    % Compute order of accuracy
    order = zeros(numel(errors) - 1, 1);
    for i = 1:numel(errors) - 1
        order(i) = log2(errors(i) / errors(i + 1));

        if order(i) - k < -0.5
            fprintf("Test FAILED for k = %d!\n", k);
            return
        end
    end
end

fprintf("Test PASSED!\n");

File: tests_C++/test5.cpp
/**
 * Poisson accuracy test
 */

#include "mole.h"
#include <iostream>

void run_test(int k, vec grid_sizes) {
  Real west = 0; // Domain's limits
  Real east = 1;

  vec errors(grid_sizes.size());

  for (int i = 0; i < grid_sizes.size(); ++i) {
    int m = grid_sizes(i);       // Number of cells
    Real dx = (east - west) / m; // Step length

    Laplacian L(k, m, dx);

    // Impose Robin BC on laplacian operator
    RobinBC BC(k, m, dx, 1, 1);
    L = L + BC;

    // 1D Staggered grid
    vec grid(m + 2);
    grid(0) = west;
    grid(1) = west + dx / 2.0;
    for (int j = 2; j <= m; j++) {
      grid(j) = grid(j - 1) + dx;
    }
    grid(m + 1) = east;

    // RHS
    vec U = exp(grid);
    U(0) = 0;              // West BC
    U(m + 1) = 2 * exp(1); // East BC

// Solve the system of linear equations
#ifdef EIGEN
    // Use Eigen only if SuperLU (faster) is not available
    vec computed_solution = Utils::spsolve_eigen(L, U);
#else
    vec computed_solution = spsolve(L, U); // Will use SuperLU
#endif

    // Compute error
    vec analytical_solution = exp(grid);
    errors(i) = max(abs(computed_solution - analytical_solution));
  }

  // Compute order of accuracy
  vec order(errors.size() - 1);
  for (int i = 0; i < errors.size() - 1; ++i) {
    order(i) = log2(errors(i) / errors(i + 1));

    if (order(i) - k < -0.5) {
      cout << "\033[1;31mTest FAILED for k = " << k << "!\033[0m\n";
      exit(1);
    }
  }
}

int main() {
  vec grid_sizes = {20, 40}; // Different grid sizes to test
  for (int k : {2, 4, 6})
    run_test(k, grid_sizes);

  cout << "\033[1;32mTest PASSED!\033[0m\n";

  return 0;
}
